semanticdb/semanticdb3/src/main/scala/scala/meta/internal/semanticdb3/Scala.scala

```diff
@@ -0,0 +1,253 @@
+    def isRootPackage: Boolean = symbol == Symbols.RootPackage
+    def isEmptyPackage: Boolean = symbol == Symbols.EmptyPackage
+    def isGlobal: Boolean = !isNone && Descriptor.descriptorLasts.contains(symbol.last)
+    def isLocal: Boolean = !isNone && !isGlobal
+    def isValid: Boolean = ??? // TODO: If we find ourselves needing this, we'll need a parser.
+    def ownerChain: List[String] = if (!owner.isNone) owner.ownerChain :+ symbol else List(symbol)
+    def owner: String = {
+      if (isGlobal) {
+        if (isRootPackage) Symbols.None
+        else {
+          val rest = DescriptorParser(symbol)._2
+          if (rest.nonEmpty) rest
+          else Symbols.RootPackage
+        }
+      } else {
+        Symbols.None
+      }
+    }
+    def desc: Descriptor = DescriptorParser(symbol)._1
+  }
+
+  sealed trait Descriptor {
+    def isNone: Boolean = this == d.None
+    def isTerm: Boolean = this.isInstanceOf[d.Term]
+    def isMethod: Boolean = this.isInstanceOf[d.Method]
+    def isType: Boolean = this.isInstanceOf[d.Type]
+    def isParameter: Boolean = this.isInstanceOf[d.Parameter]
+    def isTypeParameter: Boolean = this.isInstanceOf[d.TypeParameter]
+    def name: String
+    override def toString: String = {
+      this match {
+        case d.None => sys.error("unsupported descriptor")
+        case d.Term(name) => s"${name.encoded}."
+        case d.Method(name, tpe, 0) => s"${name.encoded}(${tpe})."
+        case d.Method(name, tpe, n) => s"${name.encoded}(${tpe}+${n})."
+        case d.Type(name) => s"${name.encoded}#"
+        case d.Parameter(name) => s"(${name.encoded})"
+        case d.TypeParameter(name) => s"[${name.encoded}]"
+      }
+    }
+  }
+  object Descriptor {
+    final case object None extends Descriptor { def name: String = n.None }
+    final case class Term(name: String) extends Descriptor
+    final case class Method(name: String, tpe: TypeDescriptor, index: Int) extends Descriptor
+    final case class Type(name: String) extends Descriptor
+    final case class Parameter(name: String) extends Descriptor
+    final case class TypeParameter(name: String) extends Descriptor
+    private[semanticdb3] val descriptorLasts = Set('.', '#', ')', ']')
+  }
+
+  implicit class ScalaNameOps(name: String) {
+    def encoded: String = {
+      if (name == n.None) {
+        n.None
+      } else {
+        val (start, parts) = (name.head, name.tail)
+        val isStartOk = Character.isJavaIdentifierStart(start)
+        val isPartsOk = parts.forall(Character.isJavaIdentifierPart)
+        if (isStartOk && isPartsOk) name
+        else "`" + name + "`"
+      }
+    }
+    def decoded: String = {
+      name.stripPrefix("`").stripSuffix("`")
+    }
+  }
+
+  object Names {
+    val None: String = ""
+    val RootPackage: String = "_root_"
+    val EmptyPackage: String = "_empty_"
+    val Constructor: String = "<init>"
+    val Anonymous: String = "_"
+  }
+
+  sealed trait TypeDescriptor {
```

Comment:
Is this level of detail necessary? I personally prefer `Method(name: String, disambiguator: String)` in langmeta.Symbol. By making the descriptor a plain string instead of `TypeDescriptor` it seems we can avoid the `?` simplification (keeping the more readable `Int*`/`=>Int`/`{}`/...).

Reply:
The old scheme also exposes quite a lot of string manipulation and magic strings to clients of the API. But I agree that the new scheme doesn't look so pretty. I'll revert and we'll think about this later.